File: 2022/day07/python/solution.py
#!/usr/bin/env python3
import os
from collections import defaultdict

def parse_filesystem(lines):
    """Parse terminal output and return directory sizes."""
    # Track current path and directory sizes
    path = []
    dir_sizes = defaultdict(int)

    for line in lines:
        if line.startswith('$ cd'):
            target = line[5:]
            if target == '/':
                path = ['/']
            elif target == '..':
                path.pop()
            else:
                path.append(target)
        elif line.startswith('$ ls'):
            continue
        elif line.startswith('dir '):
            continue
        else:
            # It's a file with size
            size = int(line.split()[0])
            # Add size to current directory and all parent directories
            for i in range(len(path)):
                dir_path = '/'.join(path[:i+1]) or '/'
                dir_sizes[dir_path] += size

    return dir_sizes

def part1(dir_sizes):
    """Sum of sizes of directories with total size <= 100000."""
    return sum(size for size in dir_sizes.values() if size <= 100000)

def part2(dir_sizes):
    """Find smallest directory to delete to free enough space."""
    total_space = 70000000
    needed_space = 30000000
    used_space = dir_sizes['/']
    free_space = total_space - used_space
    need_to_free = needed_space - free_space

    # Find smallest directory >= need_to_free
    candidates = [size for size in dir_sizes.values() if size >= need_to_free]
    return min(candidates)

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        lines = f.read().strip().split('\n')

    dir_sizes = parse_filesystem(lines)

    print('Part 1:', part1(dir_sizes))
    print('Part 2:', part2(dir_sizes))

if __name__ == '__main__':
    main()

File: 2022/day07/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseFilesystem(lines) {
  const path = [];
  const dirSizes = new Map();

  for (const line of lines) {
    if (line.startsWith('$ cd')) {
      const target = line.slice(5);
      if (target === '/') {
        path.length = 0;
        path.push('/');
      } else if (target === '..') {
        path.pop();
      } else {
        path.push(target);
      }
    } else if (line.startsWith('$ ls') || line.startsWith('dir ')) {
      continue;
    } else {
      // It's a file with size
      const size = parseInt(line.split(' ')[0], 10);
      // Add size to current directory and all parent directories
      for (let i = 0; i < path.length; i++) {
        const dirPath = path.slice(0, i + 1).join('/') || '/';
        dirSizes.set(dirPath, (dirSizes.get(dirPath) || 0) + size);
      }
    }
  }

  return dirSizes;
}

function part1(dirSizes) {
  let sum = 0;
  for (const size of dirSizes.values()) {
    if (size <= 100000) {
      sum += size;
    }
  }
  return sum;
}

function part2(dirSizes) {
  const totalSpace = 70000000;
  const neededSpace = 30000000;
  const usedSpace = dirSizes.get('/');
  const freeSpace = totalSpace - usedSpace;
  const needToFree = neededSpace - freeSpace;

  let smallest = Infinity;
  for (const size of dirSizes.values()) {
    if (size >= needToFree && size < smallest) {
      smallest = size;
    }
  }
  return smallest;
}

const inputFile = join(__dirname, '..', 'input.txt');
const lines = readFileSync(inputFile, 'utf-8').trim().split('\n');
const dirSizes = parseFilesystem(lines);

console.log('Part 1:', part1(dirSizes));
console.log('Part 2:', part2(dirSizes));
